docs/static/persistent-queues.asciidoc

```diff
@@ -1,44 +1,197 @@
+
+[[pq-config-notes]]
+==== Configuration notes
+
+Every situation and environment is different, and the "ideal" configuration varies.
+If you optimize for performance, you may increase your risk of losing data.
+If you optimize for data protection, you may impact performance. 
+
+[[pq-size]]
+===== Queue size
+You can control queue size with the `queue.max_events` and  `queue.max_bytes` settings.
+If both settings are specified, Logstash uses whichever criteria is reached
+first. 
+See <<backpressure-persistent-queue>> for behavior when queue limits are
+reached.
+
+[[pq-lower-max_bytes]]
+===== Smaller queue size
+If you are using persistent queues to protect against data loss, but don't
+require much buffering, you can set `queue.max_bytes` to a smaller value, such
+as 10mb.
+A smaller value produces smaller queues and improves queue performance.
+
+*Sample configuration*
+
+[source, yaml]
+-----
+queue.type: persisted
+queue.max_bytes: 4gb 
+-----
+
+[[pq-fewer-checkpoints]]
+===== Fewer checkpoints
+
+Setting `queue.checkpoint.writes` and `queue.checkpoint.acks` to `0` may
+yield maximum performance, but may have potential impact on durability.
+
+In a situation where Logstash is terminated or there is a hardware-level
+failure, any data that has not been checkpointed, is lost. 
+See <<durability-persistent-queues>> to better understand the trade-offs.
+
+[[troubleshooting-pqs]]
+==== Troubleshooting persistent queues
+
+If you see an error message similar to this one, you likely have a persistent queue problem.
+
+```
+message=>"java.io.IOException: Page file size is too small to hold elements
+```
+
+This error indicates that the head page (the oldest in a folder and the one with lowest page id) has a size < 18 bytes, the size of a page header.
+
+
+To resolve the issue: 
 
-The following are problems not solved by the persistent queue feature:
+. Identify the queue (or queues) that may be corrupt.
+. Stop Logstash.
+. Run `pqrepair <path>` for each of the corrupted queues.
+
+[[pqcheck]]
+===== `pqcheck` utility
+
+Run the `pqcheck` utility to identify which persistent queue--or queues--have been corrupted. 
+
+////
+//Info to add:
+* run from where? Can you run at the folder level (or pass a path) and have it search multiple queues at once? 
+* any additional parameters (either required or optional)?
+* what form does the feedback take
+* include sample? 
+
+////
+
+[source,txt]
+-----
+pqcheck 
+-----
+
+ [[pqrepair]]
+ ===== `pqrepair` utility
+
+[source,txt]
+-----
+pqrepair <queue-folder> 
+-----
+
+The `pqrepair` utility tries to remove corrupt queue segments to bring the queue back into working order. 
+
+It starts searching from the folder where is launched and looks for `data/queue/main`.
+
+NOTE: The queue may lose some data in this operation.
+
+//?? Pass the path to the folder of a queue, for example in the user context could be /var/lib/logstash/router.
+
+////
+//Info to add:
+* run from where
```

Comment:
Running from LOGSTASH_HOME, you'd do something like

```
bin/pqrepair <QUEUE_FOLDER>
```

where `QUEUE_FOLDER` is the fully qualified path to the location of the PQ

(`--help` doesn't work on `pqrepair`)

Assuming that Logstash doesn't start and/or `pqcheck` shows an anomaly, such as `NOT_FOUND` for a page, run `pqrepair` on a folder

If it runs properly, there is basically no output:

```
bin/pqrepair data/queue/main/
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
```

Note that when logstash is run as a service, folder permissions may cause an issue, as `pqrepair` requires write access to the folder, which would lead to something like:

```
ubuntu@bigger:/usr/share/logstash$ bin/pqrepair /var/lib/logstash/queue/main/
Using bundled JDK: /usr/share/logstash/jdk
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Exception in thread "main" java.nio.file.AccessDeniedException: /var/lib/logstash/queue/main/page.2
```

If so, use `sudo`:

```
ubuntu@bigger:/usr/share/logstash$ sudo -u logstash bin/pqrepair /var/lib/logstash/queue/main/
Using bundled JDK: /usr/share/logstash/jdk
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
```

Reply:
After running `pqrepair`, restart logstash to verify that the repair worked